File: src/com/valvesoftware/ValveActivity2.java
package com.valvesoftware;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.app.Activity;
import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.graphics.Point;
import android.os.Bundle;
import android.view.Display;
import java.util.HashMap;
import java.io.File;
import java.util.Locale;
import org.libsdl.app.SDLActivity;
import me.nillerusr.LauncherActivity;
import android.content.SharedPreferences;
import android.content.Context;
import android.util.Log;
import me.nillerusr.ExtractAssets;
import me.nillerusr.md3.Md3Theme;

public class ValveActivity2 { // not activity, i am lazy to change native methods
	private static Activity mSingleton;
	public static SharedPreferences mPref;


	public static native void setArgs(String args);
	public static native int setenv(String name, String value, int overwrite);
	private static native void nativeOnActivityResult(Activity activity, int i, int i2, Intent intent);

	public static int findGameinfo(String path)
	{
		File dir = new File(path);
		boolean havePlatform = false;
		boolean haveGameinfo = false;

		if( !dir.isDirectory() )
			return 0;

		for( File file : dir.listFiles() )
		{
			if( file.isDirectory() )
			{
				for( File f : file.listFiles() )
				{
					if( f.getName().toLowerCase().equals("gameinfo.txt") )
						haveGameinfo = true;
				}
			}

			if( file.getName().toLowerCase().equals("platform") )
				havePlatform = true;
		}

		if( !haveGameinfo )
			return 0;

		if( !havePlatform )
			return -1;

		return 1;
	}

	static public boolean isModGameinfoExists(String path)
	{
		File dir = new File(path);
		if( !dir.isDirectory() )
			return false;

		for( File file : dir.listFiles() )
		{
			if( file.isFile() && file.getName().toLowerCase().equals("gameinfo.txt") )
				return true;
		}

		return false;
	}

	static public int preInit(Context context, Intent intent)
	{
		mPref = context.getSharedPreferences("mod", 0);
		String gamepath = mPref.getString("gamepath", LauncherActivity.getDefaultDir() + "/srceng");
		String gamedir = intent.getStringExtra("gamedir");
		if( gamedir == null || gamedir.isEmpty() )
			gamedir = "hl2";

		int haveGameinfo = findGameinfo(gamepath);
		boolean haveModGameinfo = isModGameinfoExists(gamepath+"/"+gamedir);

		if( haveGameinfo == 0 || !haveModGameinfo )
			return 0;

		if( haveGameinfo == -1 )
			return -1;

		return 1;
	}

	static public void initNatives(Context context, Intent intent) {
		mPref = context.getSharedPreferences("mod", 0);
		ApplicationInfo appinf = context.getApplicationInfo();
		String gamepath = mPref.getString("gamepath", LauncherActivity.getDefaultDir() + "/srceng");

		String argv = intent.getStringExtra("argv");
		String gamedir = intent.getStringExtra("gamedir");
		String gamelibdir = intent.getStringExtra("gamelibdir");
		String customVPK = intent.getStringExtra("vpk");
		Log.v("SRCAPK", "argv="+argv);

		if( gamedir == null || gamedir.isEmpty() )
			gamedir = "hl2";

		if( argv == null || argv.isEmpty() )
			argv = mPref.getString("argv", "-nobackgroundlevel");

		argv = "-game "+gamedir+" "+argv;

		// Append "-language <lang>" transparently from launcher settings (MD3).
		// Empty means: don't force, let the game decide from Steam/LANG env.
		String gameLang = Md3Theme.getGameLang(context);
		if( gameLang != null && !gameLang.isEmpty() ) {
			// Sanitize: only allow [a-z_] to prevent argv injection
			if( gameLang.matches("[a-z_]+") ) {
				argv = argv + " -language " + gameLang;
			}
		}

		// Append resolution args: -w <width> -h <height> + -full / -windowed
		// DEVICE mode (0,0) means: don't add -w/-h, let engine use device native res.
		int[] res = Md3Theme.getResolvedResolution(context);
		if( res != null && res.length >= 2 && res[0] > 0 && res[1] > 0 ) {
			int w = res[0], h = res[1];
			// Sanitize: reasonable range (320..8192 for both)
			if( w >= 320 && w <= 8192 && h >= 240 && h <= 8192 ) {
				argv = argv + " -w " + w + " -h " + h;
			}
		}
		// Fullscreen / windowed
		boolean fullscreen = Md3Theme.getResolutionFullscreen(context);
		argv = fullscreen ? (argv + " -full") : (argv + " -windowed");

		if( gamelibdir != null && !gamelibdir.isEmpty() )
			setenv( "APP_MOD_LIB", gamelibdir, 1 );

		ExtractAssets.extractAssets(context);

		String vpks = context.getFilesDir().getPath()+"/"+ExtractAssets.VPK_NAME;
		if( customVPK != null && !customVPK.isEmpty() )
			vpks = customVPK+","+vpks;

		Log.v("SRCAPK", "vpks="+vpks);

		setenv( "EXTRAS_VPK_PATH", vpks, 1 );
		setenv( "LANG", Locale.getDefault().toString(), 1 );
		setenv( "APP_DATA_PATH", appinf.dataDir, 1);
		setenv( "APP_LIB_PATH", appinf.nativeLibraryDir, 1);

		if (mPref.getBoolean("rodir", false))
			setenv( "VALVE_GAME_PATH", LauncherActivity.getAndroidDataDir(), 1 );
		else
			setenv( "VALVE_GAME_PATH", gamepath, 1 );

		Log.v("SRCAPK", "argv="+argv);
		setArgs(argv);
	}
}

File: src/com/valvesoftware/ValveActivity2.kt
package com.valvesoftware

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.util.Log
import java.io.File
import java.util.Locale
import me.nillerusr.ExtractAssets
import me.nillerusr.LauncherActivity
import me.nillerusr.md3.Md3Theme

class ValveActivity2 {
    companion object {
        @JvmField
        var mPref: SharedPreferences? = null

        const val PREF_SKIP_INTRO = "skip_intro_video"

        @JvmStatic
        external fun setArgs(args: String)

        @JvmStatic
        external fun setenv(name: String, value: String, overwrite: Int): Int

        @JvmStatic
        private external fun nativeOnActivityResult(
            activity: Activity,
            requestCode: Int,
            resultCode: Int,
            intent: Intent
        )

        @JvmStatic
        fun findGameinfo(path: String): Int {
            val directory = File(path)
            if (!directory.isDirectory) return 0

            var hasPlatform = false
            var hasGameinfo = false
            directory.listFiles()?.forEach { file ->
                if (file.isDirectory) {
                    hasGameinfo = hasGameinfo || file.listFiles()?.any {
                        it.name.lowercase(Locale.ROOT) == "gameinfo.txt"
                    } == true
                }
                if (file.name.lowercase(Locale.ROOT) == "platform") hasPlatform = true
            }

            if (!hasGameinfo) return 0
            return if (hasPlatform) 1 else -1
        }

        @JvmStatic
        fun isModGameinfoExists(path: String): Boolean {
            val directory = File(path)
            if (!directory.isDirectory) return false
            return directory.listFiles()?.any {
                it.isFile && it.name.lowercase(Locale.ROOT) == "gameinfo.txt"
            } == true
        }

        @JvmStatic
        fun preInit(context: Context, intent: Intent): Int {
            val preferences = context.getSharedPreferences("mod", 0)
            mPref = preferences
            val gamePath = preferences.getString(
                "gamepath",
                LauncherActivity.getDefaultDir() + "/srceng"
            ) ?: LauncherActivity.getDefaultDir() + "/srceng"
            val gameDirectory = intent.getStringExtra("gamedir").orEmpty().ifEmpty { "hl2" }

            val gameinfo = findGameinfo(gamePath)
            if (gameinfo == 0 || !isModGameinfoExists("$gamePath/$gameDirectory")) return 0
            return if (gameinfo == -1) -1 else 1
        }

        @JvmStatic
        fun initNatives(context: Context, intent: Intent) {
            val preferences = context.getSharedPreferences("mod", 0)
            mPref = preferences
            val applicationInfo = context.applicationInfo
            val gamePath = preferences.getString(
                "gamepath",
                LauncherActivity.getDefaultDir() + "/srceng"
            ) ?: LauncherActivity.getDefaultDir() + "/srceng"
            val gameDirectory = intent.getStringExtra("gamedir").orEmpty().ifEmpty { "hl2" }
            val gameLibraryDirectory = intent.getStringExtra("gamelibdir")
            val customVpk = intent.getStringExtra("vpk")
            var arguments = intent.getStringExtra("argv").orEmpty()
            Log.v("SRCAPK", "argv=$arguments")

            if (arguments.isEmpty()) {
                arguments = preferences.getString("argv", "-nobackgroundlevel")
                    ?: "-nobackgroundlevel"
            }
            arguments = "-game $gameDirectory $arguments"

            val gameLanguage = Md3Theme.getGameLang(context)
            if (gameLanguage.matches(Regex("[a-z_]+"))) {
                arguments = arguments
                    .replace(Regex("(?i)(^|\\s)-language(?:\\s+|=)\\S+"), "$1")
                    .trim()
                arguments += " -language $gameLanguage"
            }

            if (preferences.getBoolean(PREF_SKIP_INTRO, false)) {
                arguments = arguments.replace(Regex("(?i)(^|\\s)-novid(?:\\s|$)"), "$1").trim()
                arguments += " -novid"
            }

            arguments = arguments
                .replace(Regex("(?i)(^|\\s)-(?:w|width|h|height)(?:\\s+|=)\\S+"), "$1")
                .trim()
            val resolution = Md3Theme.getResolvedResolution(context)
            if (resolution.size >= 2) {
                val width = resolution[0]
                val height = resolution[1]
                if (width in 320..8192 && height in 240..8192) {
                    arguments += " -w $width -h $height"
                }
            }

            if (!gameLibraryDirectory.isNullOrEmpty()) setenv("APP_MOD_LIB", gameLibraryDirectory, 1)

            ExtractAssets.extractAssets(context)
            var vpks = context.filesDir.path + "/" + ExtractAssets.VPK_NAME
            if (!customVpk.isNullOrEmpty()) vpks = "$customVpk,$vpks"
            Log.v("SRCAPK", "vpks=$vpks")

            setenv("EXTRAS_VPK_PATH", vpks, 1)
            setenv("LANG", Md3Theme.getRealSystemLocale().toString(), 1)
            setenv("APP_DATA_PATH", applicationInfo.dataDir, 1)
            setenv("APP_LIB_PATH", applicationInfo.nativeLibraryDir, 1)
            setenv(
                "VALVE_GAME_PATH",
                if (preferences.getBoolean("rodir", false)) LauncherActivity.getAndroidDataDir() else gamePath,
                1
            )

            Log.v("SRCAPK", "argv=$arguments")
            setArgs(arguments)
        }
    }
}
